hop/io.py

```diff
@@ -3,44 +3,218 @@
 __author__ = "Patrick Godwin (patrick.godwin@psu.edu)"
 __description__ = "a module for i/o utilities"
 
+from collections import namedtuple
+from contextlib import contextmanager
+import getpass
+from enum import Enum
+import json
+import logging
+import random
+import string
+import warnings
 
-from adc import streaming
+from adc import consumer, errors, kafka, producer
+
+from . import models
+
+logger = logging.getLogger("hop")
+
+StartPosition = consumer.ConsumerStartPosition
 
 
 class Stream(object):
     """Defines an event stream.
 
-    Sets up defaults used within the genesis client, so that when a
+    Sets up defaults used within the client so that when a
     stream connection is opened, it will use defaults specified here.
 
     Args:
-      format: the message serialization format
-      start_at: the message offset to start at
-      config: librdkafka style options, either a `dict` or a file path
+        auth: An Auth instance.
+        start_at: The message offset to start at.
+        persist: Whether to listen to new messages forever or stop
+             when EOS is received.
 
     """
 
-    def __init__(self, format=None, start_at=None, config=None):
-        self._options = {}
-        if format is not None:
-            self._options["format"] = format
-        if start_at is not None:
-            self._options["start_at"] = start_at
-        if config is not None:
-            self._options["config"] = config
+    def __init__(self, auth=None, start_at=None, persist=None):
+        self.auth = auth
+        self.start_at = start_at
+        self.persist = persist
 
-    def open(self, *args, **kwargs):
+    def open(self, url, mode="r", auth=None, start_at=None, persist=None, metadata=False):
         """Opens a connection to an event stream.
 
         Args:
-          broker_url: sets the broker URL to connect to
+            url: Sets the broker URL to connect to.
 
         Kwargs:
-          mode: read ('r') or write ('w') from the stream
-          format: the message serialization format
-          start_at: the message offset to start at
-          config: librdkafka style options, either a `dict` or a file path
+            mode: Read ('r') or write ('w') from the stream.
+            auth: An Auth instance.
+            start_at: The message offset to start at (read only).
+            persist: Whether to listen to new messages forever or stop
+                 when EOS is received (read only).
```

Comment:
I can give some motivation for why I took this approach, and then you can tell me whether or not this is a crazy idea. So, I was trying to emulate request's `Session` object (https://requests.readthedocs.io/en/master/user/advanced/) here when making the `Stream` here to persist certain parameters across open connections, but being able to override them as needed when making a call to `open()`. Part of this is a bit silly since we're not really keeping around the same Producer/Consumer objects across multiple calls to `open()`, so some of the benefits of keeping this doesn't really apply here. Nonetheless, I was hoping that some of this was handy even if it meant keeping a little extra complexity on our side. In particular, I wanted to support two ends of use cases while making it kinda nice and flexible to do both. 

In the first, I only care about making a single connection and don't care about just putting everything into the `open()` call. In that case, carrying around the burden of the `Stream` class is a little much and I added a convenience wrapper, `stream`, which maps directly to `Stream()`. As a result, I can do something like this:

```
from hop import stream

with stream.open("kafka://myhostname/topic", "r", persist=True) as s:
    for message in s:
        print(s)
```

In the second case, I would opening up multiple producers and/or consumers and may want to save some common configuration ahead of time, such as providing auth credentials. Say, reading from a topic and writing the same message to another topic. Then I could do something like:

```
from hop import Stream
from hop.auth import load_auth

s = Stream(auth=load_auth(), persist=True)
with s.open("kafka://myhostname/topic1", "r") as source, s.open("kafka://myhostname/topic2", "w") as sink:
    for message in source:
        sink.write(message)
```

Reply:
Okay, I understand.

I don't think that's the API I would design - you save on typing `auto=load_auth(), persist=True` twice, but there isn't much other benefit - but it makes sense. As long as you're thinking about it, I'd say you should use your judgement.